File: tests/tests_mashnetwork/echo_client.py
#! /usr/bin/env python

################################################################################
# The MASH Framework contains the source code of all the servers in the
# "computation farm" of the MASH project (http://www.mash-project.eu),
# developed at the Idiap Research Institute (http://www.idiap.ch).
#
# Copyright (c) 2016 Idiap Research Institute, http://www.idiap.ch/
# Written by Philip Abbet (philip.abbet@idiap.ch)
#
# This file is part of the MASH Framework.
#
# The MASH Framework is free software: you can redistribute it and/or modify
# it under the terms of either the GNU General Public License version 2 or
# the GNU General Public License version 3 as published by the Free
# Software Foundation, whichever suits the most your needs.
#
# The MASH Framework is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public Licenses
# along with the MASH Framework. If not, see <http://www.gnu.org/licenses/>.
################################################################################


import sys
import os


def CHECK_EQUAL(expected, actual):
    if expected != actual:
        print "Failure: expected %s, got %s" % (str(expected), str(actual))
        sys.exit(-1)



def testTransmission(client, command):
    if not(client.sendCommand(command)):
        print "Failed to send command '%s' to the server" % command
        return None

    response = client.waitResponse()
    if response is None:
        print "Failed to wait for response to command '%s' from the server" % command.toString()
        return None

    CHECK_EQUAL(command.toString(), response.toString());

    return response



# Parameters handling
if (len(sys.argv) > 2) or ((len(sys.argv) == 2) and (sys.argv[1] == '--help')):
    print "Usage: %s [MODULES_PATH]" % sys.argv[0]
    sys.exit(-1)

if len(sys.argv) == 2:
    sys.path.append(sys.argv[1])

from pymash import Client
from pymash import Message
    

# Create the client and connect to the server
client = Client()
if not(client.connect('127.0.0.1', 10000)):
    print 'Failed to connect to the server'
    sys.exit(-1)


response = testTransmission(client, Message("NO_ARGUMENT"))
if response is None:
    sys.exit(-1)


commandArgs = ["http://something.com"]
response = testTransmission(client, Message("ONE_STRING_ARGUMENT", commandArgs))
if response is None:
    sys.exit(-1)

CHECK_EQUAL(commandArgs[0], response.parameters[0]);


commandArgs = [100]
response = testTransmission(client, Message("ONE_INT_ARGUMENT", commandArgs))
if response is None:
    sys.exit(-1)

CHECK_EQUAL(commandArgs[0], int(response.parameters[0]));


commandArgs = [0.8]
response = testTransmission(client, Message("ONE_FLOAT_ARGUMENT", commandArgs))
if response is None:
    sys.exit(-1)

CHECK_EQUAL(commandArgs[0], float(response.parameters[0]));


commandArgs = ["http://something.com", 100, 0.8]
response = testTransmission(client, Message("THREE_ARGUMENTS", commandArgs))
if response is None:
    sys.exit(-1)

CHECK_EQUAL(commandArgs[0], response.parameters[0]);
CHECK_EQUAL(commandArgs[1], int(response.parameters[1]));
CHECK_EQUAL(commandArgs[2], float(response.parameters[2]));


commandArgs = ["Hello World!"]
response = testTransmission(client, Message("ONE_QUOTED_STRING_ARGUMENT", commandArgs))
if response is None:
    sys.exit(-1)

CHECK_EQUAL(commandArgs[0], response.parameters[0]);


commandArgs = ["Escaped 'Hello world!' string"]
response = testTransmission(client, Message("ONE_QUOTED_STRING_WITH_ESCAPING_ARGUMENT", commandArgs))
if response is None:
    sys.exit(-1)

CHECK_EQUAL(commandArgs[0], response.parameters[0]);


commandArgs = ["Long string on\nseveral lines"]
response = testTransmission(client, Message("ONE_QUOTED_STRING_WITH_SEVERAL_LINES_ARGUMENT", commandArgs))
if response is None:
    sys.exit(-1)

CHECK_EQUAL(commandArgs[0], response.parameters[0]);


client.close()

File: tests/tests_mashnetwork/echo_client.cpp
#include <mash-network/client.h>
#include <tests.h>
#include <iostream>

using namespace Mash;
using namespace std;


int testTransmission(Client* pClient, const std::string& strCommand, const ArgumentsList& commandArgs, ArgumentsList* responseArgs)
{
    string strResponse;

    if (!pClient->sendCommand(strCommand, commandArgs))
    {
        cerr << "Failed to send command '" << strCommand << "' to the server" << endl;
        return -1;
    }

    if (!pClient->waitResponse(&strResponse, responseArgs))
    {
        cerr << "Failed to wait for response to command '" << strCommand << "' from the server" << endl;
        return -1;
    }

    CHECK_EQUAL(strCommand, strResponse);
    CHECK_EQUAL(commandArgs.size(), responseArgs->size());

    return 0;
}



int main(int argc, char** argv)
{
    // Declarations
    Client client;
    ArgumentsList commandArgs;
    ArgumentsList responseArgs;
    
    
    if (!client.connect("127.0.0.1", 10000))
    {
        cerr << "Failed to connect to the server" << endl;
        return -1;
    }
    

    commandArgs.clear();
    
    if (testTransmission(&client, "NO_ARGUMENT", commandArgs, &responseArgs) != 0)
        return -1;
    

    commandArgs.clear();
    commandArgs.add("http://something.com");

    if (testTransmission(&client, "ONE_STRING_ARGUMENT", commandArgs, &responseArgs) != 0)
        return -1;

    CHECK_EQUAL(commandArgs.getString(0), responseArgs.getString(0));

    
    commandArgs.clear();
    commandArgs.add(100);

    if (testTransmission(&client, "ONE_INT_ARGUMENT", commandArgs, &responseArgs) != 0)
        return -1;

    CHECK_EQUAL(commandArgs.getInt(0), responseArgs.getInt(0));


    commandArgs.clear();
    commandArgs.add(0.8f);

    if (testTransmission(&client, "ONE_FLOAT_ARGUMENT", commandArgs, &responseArgs) != 0)
        return -1;

    CHECK_EQUAL(commandArgs.getFloat(0), responseArgs.getFloat(0));


    commandArgs.clear();
    commandArgs.add("http://something.com");
    commandArgs.add(100);
    commandArgs.add(0.8f);

    if (testTransmission(&client, "THREE_ARGUMENTS", commandArgs, &responseArgs) != 0)
        return -1;

    CHECK_EQUAL(commandArgs.getString(0), responseArgs.getString(0));
    CHECK_EQUAL(commandArgs.getInt(1), responseArgs.getInt(1));
    CHECK_EQUAL(commandArgs.getFloat(2), responseArgs.getFloat(2));


    commandArgs.clear();
    commandArgs.add("Hello world!");

    if (testTransmission(&client, "ONE_QUOTED_STRING_ARGUMENT", commandArgs, &responseArgs) != 0)
        return -1;

    CHECK_EQUAL(commandArgs.getString(0), responseArgs.getString(0));


    commandArgs.clear();
    commandArgs.add("Escaped 'Hello world!' string");

    if (testTransmission(&client, "ONE_QUOTED_STRING_WITH_ESCAPING_ARGUMENT", commandArgs, &responseArgs) != 0)
        return -1;

    CHECK_EQUAL(commandArgs.getString(0), responseArgs.getString(0));


    commandArgs.clear();
    commandArgs.add("Long string on\nseveral lines");

    if (testTransmission(&client, "ONE_QUOTED_STRING_WITH_SEVERAL_LINES_ARGUMENT", commandArgs, &responseArgs) != 0)
        return -1;

    CHECK_EQUAL(commandArgs.getString(0), responseArgs.getString(0));


    client.close();
    
    return 0;
}
